chia/wallet/util/generate_curry_and_treehash_f.py

```diff
@@ -0,0 +1,72 @@
+from hashlib import sha256
+from typing import Callable, List
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+
+CurryHashFunction = Callable[..., bytes32]
+
+
+NULL = bytes.fromhex("")
+ONE = bytes.fromhex("01")
+TWO = bytes.fromhex("02")
+Q_KW = bytes.fromhex("01")
+A_KW = bytes.fromhex("02")
+C_KW = bytes.fromhex("04")
+
+
+def shatree_atom(atom: bytes) -> bytes32:
+    s = sha256()
+    s.update(ONE)
+    s.update(atom)
+    return bytes32(s.digest())
+
+
+def shatree_pair(left_hash: bytes32, right_hash: bytes32) -> bytes32:
+    s = sha256()
+    s.update(TWO)
+    s.update(left_hash)
+    s.update(right_hash)
+    return bytes32(s.digest())
+
+
+Q_KW_TREEHASH = shatree_atom(Q_KW)
+A_KW_TREEHASH = shatree_atom(A_KW)
+C_KW_TREEHASH = shatree_atom(C_KW)
+ONE_TREEHASH = shatree_atom(ONE)
+NULL_TREEHASH = shatree_atom(NULL)
+
+
+# The environment `E = (F . R)` recursively expands out to
+# `(c . ((q . F) . EXPANSION(R)))` if R is not 0
+# `1` if R is 0
+
+
+def curried_values_tree_hash(arguments: List[bytes32]) -> bytes32:
+    if len(arguments) == 0:
+        return ONE_TREEHASH
+
+    return shatree_pair(
+        C_KW_TREEHASH,
+        shatree_pair(
+            shatree_pair(Q_KW_TREEHASH, arguments[0]),
+            shatree_pair(curried_values_tree_hash(arguments[1:]), NULL_TREEHASH),
+        ),
+    )
+
+
+# The curry pattern is `(a . ((q . F)  . (E . 0)))` == `(a (q . F) E)
+# where `F` is the `mod` and `E` is the curried environment
+
+
+def generate_curry_and_treehash_f(mod_hash: bytes32) -> CurryHashFunction:
```

Comment:
the `_f` suffix is meant to indicate it returns a function?

Reply:
Yes, exactly.

Reply:
It might actually be overkill... we could probably just export the inner function and have `mod_hash` be a parameter, and assume the caller to be smart enough to pre-cache it where appropriate. It would probably make it easier to understand for the casual observer too. Do you have an opinion on this?

Reply:
I have a weak preference for exporting the inner function directly. I think it's simpler and more straight-forward. I'm pretty sure all wallets do cache the tree-hash of their puzzle already as well.

Reply:
I see, it's not the treehash of the puzzle directly, it's the hash of the quoted puzzle

Reply:
Right, so that's a detail that the caller would have to deal with. It would have to be clearly documented.